tests/test_generate_reports.py

```diff
@@ -0,0 +1,357 @@
+
+    expected_balance_dict = {
+        "start_time": datetime(1970, 1, 1, 0, 0, tzinfo=tzutc()),
+        "end_time": datetime(2100, 1, 1, 0, 0, tzinfo=tzutc()),
+        "total_income": total_income,
+        "total_expense": total_expense,
+        "total_balance": total_balance,
+    }
+
+    return expected_balance_dict
+
+
+def _expected_group_by_ref(grouped_transactions: List[GroupedTransaction]) -> List[Any]:
+    grouped_by_ref: Dict[Any, Any] = {}
+
+    for transaction in grouped_transactions:
+        group_number = transaction["groupNumber"]
+        grouped = grouped_by_ref.get(group_number)
+        if grouped is None:
+            grouped_by_ref[group_number] = {
+                "key": group_number,
+                "amount": transaction["amount"],
+                "references": [transaction["referenceText"]],
+                "groupName": transaction["groupName"],
+            }
+        else:
+            grouped["amount"] += transaction["amount"]
+            grouped["references"] += [transaction["referenceText"]]
+
+    grouped_by_ref_list = list(grouped_by_ref.values())
+    return grouped_by_ref_list
+
+
+def _expected_categ_group(
+    grouped_transactions: List[GroupedTransaction],
+    group_references: List[set[str]],
+) -> List[Any]:
+    transactions = deepcopy(grouped_transactions)
+    transactions_ref = deepcopy(group_references)
+    categories_group: Dict[Any, Any] = {}
+
+    for transaction in transactions:
+        key_category = get_category(
+            transaction["referenceText"],
+            transactions_ref[transaction["groupNumber"]],
+            fallback_reference=transaction["groupName"],
+        )
+        category_group = categories_group.get(key_category)
+        if category_group is None:
+            categories_group[key_category] = {
+                "key": key_category,
+                "amount": transaction["amount"],
+                "references": [transaction["referenceText"]],
+            }
+        else:
+            category_group["amount"] += transaction["amount"]
+            category_group["references"] += [transaction["referenceText"]]
+
+    categories_group_list = list(categories_group.values())
+    return categories_group_list
+
+
+def _expected_categ_transactions(
+    grouped_transactions: List[GroupedTransaction],
+    group_references: List[set[str]],
+) -> List[Any]:
+
+    categorized_transactions: List[Any] = deepcopy(grouped_transactions)
+    categorized_transactions_ref = deepcopy(group_references)
+
+    for transaction in categorized_transactions:
+        key_category = get_category(
+            transaction["referenceText"],
+            categorized_transactions_ref[transaction["groupNumber"]],
+            fallback_reference=transaction["groupName"],
+        )
+        transaction["customCategory"] = key_category
+
+    return categorized_transactions
+
+
+def test_generate_reports_output(write_json_mock: Mock) -> None:
```

Comment:
Let's get rid of duplicating the business logic in unit tests and create some hard coded "expected files" that should come out of `generate_reports` when given inputs `test_transaction.json` and `test_config_file.yaml`.

It's important that we validate manually that the hard coded expected files are correct, this is a one off validation to build the test.